### ISSUE TITLE
Grouping/sorting of dependencies in Dependency Dashboard

### ISSUE THREAD

**Issue by karfau** [renovatebot/renovate#5441]
**What would you like Renovate to be able to do?**
When configuring master issue it would be nice to see what file(s) a certain PR is targeting before/without creating the PR.

To consider:
While it is "obvious" when configuring `commitMessageSuffix`, the information about configured `branchPrefix`es is not visible in the masterIssue.

**Describe the solution you'd like**

It would be nice to have a way to either group the checkboxes by targeted file(s) ~and/or `branchPrefix`~.

I'm considering to work on this feature if it is wanted.

**Describe alternatives you've considered**
- I'm not sure how the sorting is done in the master issue, but if it would at least not mix up dependencies from different ~packageRules/~ targeted file(s) ~/branchPrefix/commitMessageSuffix~ the current way of adding commitMessageSuffix might already be sufficient.

- Since users might have more ideas what to group the entries by, it might be worth wile to support something like `masterIssueGrouping` that can be set to a list of predefined keys. (no longer sure this makes sense, but I'll leave it here as an idea.)

**Additional context**

**Comment by rarkins** [comment]
I'm not sure I understand you completely, so what do you think about the following?

1. Copy/paste the raw markdown in your current issue into a comment here (the one you're not happy with)
2. Manually mock up what you think it should look like according to your idea. i.e. modify the markdown in a separate comment to match what the final look should be

**Comment by karfau** [comment]
An example of mixed things can be found here:
https://github.com/karfau/renovate-config-help-519/issues/2
(That repo is readonly now, so no need to copy it's master issue markdown here.)

I will add some comments of how I can image it could look like.
(I have no example at hand that has multiple sections but I think the groupings should happen per section in the same way.)

**Comment by karfau** [comment]
An example without `branchPrefix`:

## Section
### Changes to `src-react/package.json`

 - [ ] chore(deps): pin dependencies (`@babel/preset-typescript`, `@types/core-js`, `@types/enzyme`, `@types/jest`, `@types/react`, `@types/react-dom`, `@types/react-measure`, `@types/react-redux`, `@types/redux`, `@types/redux-saga`, `@types/styled-components`, `enzyme`, `enzyme-adapter-react-16`, `got`, `grunt`, `grunt-angular-architecture-graph`, `grunt-bump`, `grunt-coffeelint`, `grunt-contrib-clean`, `grunt-contrib-coffee`, `grunt-contrib-concat`, `grunt-contrib-copy`, `grunt-contrib-jshint`, `grunt-contrib-less`, `grunt-contrib-watch`, `grunt-conventional-changelog`, `grunt-istanbul`, `grunt-istanbul-coverage`, `grunt-karma`, `grunt-ngdocs`, `jest`, `karma-chrome-launcher`, `karma-coffee-preprocessor`, `karma-coverage`, `karma-firefox-launcher`, `karma-jasmine`, `karma-junit-reporter`, `karma-phantomjs-launcher`, `karma-requirejs`, `karma-script-launcher`, `karma-spec-reporter`, `load-grunt-configs`, `lodash`, `moment`, `phantomjs-prebuilt`, `react-test-renderer`, `request`, `requirejs`, `resolve-url-loader`, `source-map-loader`, `testingbot-api`, `time-grunt`, `ts-loader`, `typescript`, `webpack`, `webpack-cli`)
 - [ ] chore(deps): update dependency @types/enzyme to v3.10.5
 - [ ] chore(deps): update dependency @types/jest to v25.1.2
 - [ ] chore(deps): update dependency webpack to v4.41.6

### Changes to `src-react/.nvmrc`
 - [ ] chore(deps): update node.js to 8.17

### Changes to `.nvmrc` *(only after approval)*
 - *File is up to date*

### Changes to `package.json` *(only after approval)*
 - [ ] chore(deps): update dependency grunt to v1.0.4 [+(package.json)]
 - [ ] chore(deps): update dependency grunt-contrib-coffee to v2.1.0 [+(package.json)]
 - [ ] chore(deps): update dependency grunt-contrib-uglify to v2.3.0 [+(package.json)]
 - [ ] chore(deps): update dependency grunt-contrib-watch to v1.1.0 [+(package.json)]
 - [ ] chore(deps): update dependency grunt-html2js to v0.7.1 [+(package.json)]
 - [ ] chore(deps): update dependency grunt-istanbul to v0.8.0 [+(package.json)]
 - [ ] chore(deps): update dependency grunt-ngdocs to v0.2.11 [+(package.json)]
 - [ ] chore(deps): update dependency karma to v1.7.1 [+(package.json)]
 - [ ] chore(deps): update dependency karma-coverage to v1.1.2 [+(package.json)]
 - [ ] chore(deps): update dependency karma-firefox-launcher to v1.3.0 [+(package.json)]
 - [ ] chore(deps): update dependency karma-jasmine to v1.1.2 [+(package.json)]
 - [ ] chore(deps): update dependency moment to v2.24.0 [+(package.json)]
 - [ ] chore(deps): update dependency request to v2.88.0 [+(package.json)]
 - [ ] chore(deps): update dependency requirejs to v2.3.6 [+(package.json)]
 - [ ] chore(deps): update dependency testingbot-api to v1.0.7 [+(package.json)]
 - [ ] chore(deps): update dependency got to v10 [+(package.json)]
 - [ ] chore(deps): update dependency got to v7 [+(package.json)]
 - [ ] chore(deps): update dependency got to v8 [+(package.json)]
 - [ ] chore(deps): update dependency got to v9 [+(package.json)]
 - [ ] chore(deps): update dependency grunt-concurrent to v3 [+(package.json)]
 - [ ] chore(deps): update dependency grunt-contrib-clean to v2 [+(package.json)]
 - [ ] chore(deps): update dependency grunt-contrib-jshint to v2 [+(package.json)]
 - [ ] chore(deps): update dependency grunt-contrib-less to v2 [+(package.json)]
 - [ ] chore(deps): update dependency grunt-contrib-uglify to v3 [+(package.json)]
 - [ ] chore(deps): update dependency grunt-contrib-uglify to v4 [+(package.json)]
 - [ ] chore(deps): update dependency grunt-karma to v3 [+(package.json)]
 - [ ] chore(deps): update dependency grunt-protractor-runner to v5 [+(package.json)]
 - [ ] chore(deps): update dependency karma to v2 [+(package.json)]
 - [ ] chore(deps): update dependency karma to v3 [+(package.json)]
 - [ ] chore(deps): update dependency karma to v4 [+(package.json)]
 - [ ] chore(deps): update dependency karma-chrome-launcher to v3 [+(package.json)]
 - [ ] chore(deps): update dependency karma-coverage to v2 [+(package.json)]
 - [ ] chore(deps): update dependency karma-jasmine to v2 [+(package.json)]
 - [ ] chore(deps): update dependency karma-jasmine to v3 [+(package.json)]
 - [ ] chore(deps): update dependency karma-junit-reporter to v2 [+(package.json)]

**Comment by karfau** [comment]
Actually when thinking about it, the branchPrefix part is less relevant for this issue than I thought initially.

I updated both last comments multiple time during drfting the second one, sorry if that caused any confusion.

**Comment by rarkins** [comment]
This is more complex than you may anticipate, for example:

It's very common for Renovate to update multiple package files per-Pull Request. In fact it's usually desirable for multiple package files to be updated in the same PR instead of having different versions spread across projects. So the base assumption that you have (PRs are separated per-file) does not hold for most users.

What we could maybe do though is add an optional field like "masterIssueSection" to config that would let you explicitly configure 2 or more sections with the master issue. We'd need to think how it could possibly be handled if a PR contains multiple updates and you configured different sections for each - should Renovate pick one (essentially at random) or duplicate the PR amongst all sections?

Also, in the example screenshot you are looking at a simplified view where all PRs showed up as "Open". The master issue of today can already have up to 5-6 different categories. e.g. Renovate's repo has these three currently:

![image](https://user-images.githubusercontent.com/6311784/74250776-698b8c00-4ceb-11ea-8896-2fb782865298.png)

**Comment by karfau** [comment]
I think it would be good enough if they are just grouped by files that are potentially affected, not which files will actually be affected. If this helps.

In the normal case you describe it would lead to only one "group", which could just be listed once before all sections:

> This [master issue](https://renovatebot.com/blog/master-issue) contains a list of Renovate updates and their statuses.
> 
> *The following files are currently maintained:*
> - `Dockerfile`
> - `package.json`
> - `package-lock.json`
> - `package.json`
> - ...
> 
> ## Pending Approval
> These branches will be created by Renovate only once you click their checkbox below.
> 
> * [ ]  build(deps): update dependency semver to v7.1.2 (`semver`, `@types/semver`)
> * [ ]  build(deps): update dependency @sindresorhus/is to v2
> * [ ]  build(deps): update dependency azure-devops-node-api to v10
> * [ ]  build(deps): update dependency cacache to v14
> * [ ]  build(deps): update dependency chalk to v3
> * [ ]  build(deps): update dependency got to v10
> * [ ]  build(deps): update dependency markdown-table to v2
> * [ ]  build(deps): update dependency pnpm to v4
> * [ ]  build(deps): update dependency semantic-release to v17
> * [ ]  chore(deps): update dependency aws-sdk-mock to v5
> * [ ]  chore(deps): update dependency eslint-config-airbnb-typescript to v7
> * [ ]  chore(deps): update dependency husky to v4
> * [ ]  chore(deps): update dependency mock to v4
> * [ ]  chore(deps): lock file maintenance
> 
> ## Pending Status Checks
> These updates await pending status checks. To force their creation now, check the box below.
> *(no need to list or group by files, but branches could be linked)*
> 
> * [ ]  build(deps): update dependency aws-sdk to v2.615.0
> * [ ]  chore(deps): update dependency eslint-config-airbnb-typescript to v6.3.2
> * [ ]  chore(deps): update dependency nock to v11.8.2
> * [ ]  chore(deps): update dependency rimraf to v3.0.2
> * [ ]  chore(deps): update typescript-eslint monorepo to v2.19.2 (`@typescript-eslint/eslint-plugin`, `@typescript-eslint/parser`)
> 
> ## Closed/Ignored
> These updates were closed unmerged and will not be recreated unless you click a checkbox below.
> Affected files: `Dockerfile`[, `/path/to/Dockerfile`, ...] 
> *(no need to list files since the PRs are linked)*
> 
> * [ ]  [chore(deps): update amd64/ubuntu docker tag to v18.10](../pull/4746)

I think adding another configuration (`masterIssueGroup`) that is unrelated to the information that is already in the configuration is not such a good idea, but that's just my, maybe naive, opinion.

**Comment by karfau** [comment]
I would also say that sections for things that already have branches, don't need to be grouped, since the affected files can easily be looked up, which would be more precise then the grouping. (Maybe the branches could even be linked.)

**Comment by rarkins** [comment]
Closing this as it's been open for nearly 1.5 years as low priority help-wanted and it's not something we intend to do in the core team.
